cEP-0002.md

```diff
@@ -0,0 +1,403 @@
+| Status       | Proposed                                |
+| Type         | Process                                 |
+
+# Abstract
+
+This document describes how the next generation bears will be designed and
+implemented.
+
+This cEP describes a new implementation of bears in coala which should make
+ them simpler and make their design more intuitive and better.
+
+# Introduction
+
+Currently the bear design is clunky and bad. The current design makes global
+bears dependant on local bears through it's queuing mechanism. It uses
+communication signals over a queue to indicate that local bears are done with
+their execution and global bears can begin. The queueing mechanism doesn't work
+very well as it sometimes has synchronization problems. Also if local bears
+fail to run it stalls the global bears. Also we have a permanent logger thread.
+An extra logger thread consumes time and resources of the system which we can
+avoid.
+
+The new proposal aims to eliminate the use of the queuing control signals.
+Utilising asynchronous I/O and process pool to generate results from  the
+bears. No longer will the local bears be dependent on global bears to finish.
+The logging will also be handled by the main thread itself instead of
+utilising an extra thread.
+
+# Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. A `ProcessPoolExecutor` is created with the maximum number of processes
+your CPU can run in parallel.
+2. We proceed with getting an asynchronous event loop. This asynchronous event
+loop creates `Concurrent.Futures` for tasks and schedules them to the process
+pool.
+3. Each task is given a callback when it has a result ready. This callback is responsible for processing the result accordingly.
+3. The `Bear` class will get two new methods. `execute_task()` and
+`generate_tasks()`.
+    * The `generate_tasks()` is resposible for generating the correct
+    arguments for the upcoming tasks.
+    * The `execute_task()` method is scheduled by the event loop to the
+    process pool using the arguments from the `generate_tasks()` method.
+3. Each bear will generate the following type of tasks:
+    * Global bears fire tasks that process all files.
+    * Local bears fire tasks for each individual file.
+    * Arbitrary tasks such as dependency bears or other tasks are also
+    possible.
+4. A bear that has dependencies will only be scheduled to the process pool once all its dependencies have been resolved. It's done in the following way:
+    * It's checked if the bear has some dependencies.
+    * If they haven't been resolved the bear is not scheduled.
+    * As soon as the dependencies are resolved for a bear it is removed from the dependency dictionary and scheduled to the `ProcessPoolExecutor`.
+5. A special purpose `Dependencies` class will be introduced to do the above
+mentioned task. This will be responsible for keeping track of dependencies.
+Whenever a bear completes it will be removed from the dependency class.
```

Comment:
It is a bit different to that. Instead of just putting the bear at the head of the queue if the dependencies are not satisfied it only schedules a bear when the dependencies get satisfied.